File: 019/main.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        dummy = ListNode(-1)
        dummy.next = head
        fast, slow = head, dummy
        for _ in range(n):
            fast = fast.next
        while fast:
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next
        return dummy.next

import os
if os.getenv('LZS'):
    dummy = ListNode(-1)
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    dummy.next = head
    s = Solution()
    head = s.removeNthFromEnd(head, 3)
    while head:
        print(head.val)
        head = head.next

File: 019/main.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  const dummy = new ListNode(-1)
  dummy.next = head
  const points = [dummy]
  let cur = head
  while (cur) {
    points.push(cur)
    cur = cur.next
  }
  const prev = points[points.length - n - 1]
  prev.next = prev.next.next
  return dummy.next
}

if (process.env.LZS) { // local test
  function ListNode (val) {
      this.val = val
      this.next = null
  }
  const assert = require('chai').assert
  assert.deepStrictEqual(removeNthFromEnd({
    val: 1,
    next: {
      val: 2,
      next: {
        val: 3,
        next: null
      }
    }
  }, 3), {
    val: 2,
    next: {
      val: 3,
      next: null
    }
  })
}
